### PR TITLE
Add support for Razer PWM PC Fan Controller

### PR DESCRIPTION
**Pull request by wlkmanist**
I am on Win8.1, so I can't use Synapse 3 (latest versions that supports this controller) and can't write a plugin for FanControl (V63 is the latest working version for me, with LHM  lib 0.8.7). But I can backport this code to 0.8.7 or use it as is.
Controls works in full range (from 0% to 100%). Also already fixed issue with fan hotplug.

image

image

### CODE REVIEW

**Comment by wlkmanist** [comment]
Yea, now its works fine, but I can't test it on newer FanControl versions.

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+        }
+    }
+
+    private sealed class SequenceCounter
+    {
+        private byte _sequenceId = 0x00;
+
+        public byte Next()
+        {
+            do
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:406]: To normal while.

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+using System;
+using System.Threading;
+using HidSharp;
+
+namespace LibreHardwareMonitor.Hardware.Controller.Razer;
+
+internal sealed class RazerFanController : Hardware
+{
+    private HidStream _stream;
+    private readonly HidDevice _device;
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:15]: Can you reformat this document?

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+            {
+                SequenceNumber = _sequenceCounter.Next(),
+                DataLength = 0,
+                CommandClass = CommandClass.Info,
+                Command = 0x87,
+            };
+
+            if (RazerGuard.WaitRazerMutex(250))
+            {
+                do
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:41]: Can be normal while.

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+            @throw = true;
+        }
+
+        if (@throw)
+        {
+            throw new InvalidOperationException("The device is not ready.");
+        }
+    }
+
+    private Packet __WriteAndRead(Packet packet)
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:188]: Naming.., no __

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        ThrowIfNotReady();
+        _stream?.SetFeature(buffer, 0, buffer.Length);
+        Thread.Sleep(DEVICE_READ_DELAY_MS);
+        ThrowIfNotReady();
+        _stream?.GetFeature(response, 0, response.Length);
+        Packet readPacket = Packet.FromBuffer(response);
+
+        if (readPacket.Status == DeviceStatus.Busy)
+        {
+            var cts = new CancellationTokenSource(DEVICE_READ_TIMEOUT_MS);
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:202]: Switch to Stopwatch perhaps?

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+        int devTimeout = 400;
+        int devReconnectTimeout = 3000;
+
+        do
+        {
+            try
+            {
+                readPacket = __WriteAndRead(packet);
+            }
+            catch (System.IO.IOException) // Unexpected device disconnect or fan plug/unplug
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:228]: Include System.IO :)

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+            {
+                if (devTimeout <= 0)
+                {
+                    do
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:232]: Can also be normal while.

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+            _rpmSensors[i].Value = GetChannelSpeed(i);
+            _pwmControls[i].Value = _pwm[i];
+        }
+
+        RazerGuard.ReleaseRazerMutex();
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+
+    private const int DEFAULT_SPEED_CHANNEL_POWER = 50;
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerFanController.cs:294]: Part of reformatting, but all these consts can just be at top of the file.

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerGroup.cs [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using HidSharp;
+
+namespace LibreHardwareMonitor.Hardware.Controller.Razer;
+
+internal class RazerGroup : IGroup
+{
+    private readonly List<IHardware> _hardware = new();
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerGroup.cs:18]: Also reformat.

**Review by PhyxionNL — CHANGES_REQUESTED** [review]
Some small things, thanks 😎

**Comment by wlkmanist** [comment]
I don't know, how to hide the device if it disconnected. Is it possible?

**Comment by PhyxionNL** [comment]
> I don't know, how to hide the device if it disconnected. Is it possible?

You can choose not to add the device in the group when it's disconnected. For dynamic add/remove you need to implement IHardwareChanged, see an example:
https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Network/NetworkGroup.cs

For HidSharp I think you can use their Changed event:
https://github.com/IntergatedCircuits/HidSharp/blob/master/HidSharp/DeviceList.cs#L40

File: LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerGroup.cs (hunk 9 above)
Comment (PhyxionNL) [reply LibreHardwareMonitorLib/Hardware/Controller/Razer/RazerGroup.cs:14]: No, there's no new line after break statement on L34 for example.

File: LibreHardwareMonitorLib/Hardware/Mutexes.cs [hunk 10]
```diff
@@ -8,6 +8,7 @@ internal static class Mutexes
     private static Mutex _ecMutex;
     private static Mutex _isaBusMutex;
     private static Mutex _pciBusMutex;
+    private static Mutex _razerDeviceMutex;
```
Comment (PhyxionNL) [inline LibreHardwareMonitorLib/Hardware/Mutexes.cs:11]: Rename _razorDeviceMutex to _razorMutex.

**Comment by wlkmanist** [comment]
> > I don't know, how to hide the device if it disconnected. Is it possible?
> 
> You can choose not to add the device in the group when it's disconnected. For dynamic add/remove you need to implement IHardwareChanged, see an example: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Network/NetworkGroup.cs
> 
> For HidSharp I think you can use their Changed event: https://github.com/IntergatedCircuits/HidSharp/blob/master/HidSharp/DeviceList.cs#L40

Ok, mayby I will try to add that

**Comment by PhyxionNL** [comment]
> > > I don't know, how to hide the device if it disconnected. Is it possible?
> > 
> > 
> > You can choose not to add the device in the group when it's disconnected. For dynamic add/remove you need to implement IHardwareChanged, see an example: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Network/NetworkGroup.cs
> > For HidSharp I think you can use their Changed event: https://github.com/IntergatedCircuits/HidSharp/blob/master/HidSharp/DeviceList.cs#L40
> 
> Ok, maybe I will try to add that

Do you still want to do this?

**Comment by wlkmanist** [comment]
> > > > I don't know, how to hide the device if it disconnected. Is it possible?
> > > 
> > > 
> > > You can choose not to add the device in the group when it's disconnected. For dynamic add/remove you need to implement IHardwareChanged, see an example: https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/master/LibreHardwareMonitorLib/Hardware/Network/NetworkGroup.cs
> > > For HidSharp I think you can use their Changed event: https://github.com/IntergatedCircuits/HidSharp/blob/master/HidSharp/DeviceList.cs#L40
> > 
> > 
> > Ok, maybe I will try to add that
> 
> Do you still want to do this?

Probably not, or much later, maybe with another PR. This is not a device that user must plug/unplug while pc is enabled.

**Pull request merged**